From 50157b43bdc16cda029c8c3c2ef5f50640ba6833 Mon Sep 17 00:00:00 2001
From: Alan Winkowski <walan@marvell.com>
Date: Sun, 6 Jan 2019 16:09:52 +0200
Subject: [PATCH 0841/1051] net: mvpp2: fix in parser sram shift function

This patch fixes an issue when setting the parser sram shift
information. In case the shift was negative, the shift sign
bit was set but then was overridden when setting the shift value

This fix should solve issues seen in IPv6, since the negative
shift is only used for IPv6 entries.

Change-Id: Ia56b7610b09fa09fae7bac1443bf17b8dc23e839
Signed-off-by: Alan Winkowski <walan@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/1568
Tested-by: sa_ip-sw-jenkins
Tested-by: Stefan Chulski <Stefan.Chulski@cavium.com>
Reviewed-by: Stefan Chulski <Stefan.Chulski@cavium.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/ethernet/marvell/mvpp2/mvpp2_prs.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/marvell/mvpp2/mvpp2_prs.c b/drivers/net/ethernet/marvell/mvpp2/mvpp2_prs.c
index d84eaa4dc80f..8a79b3c3372d 100644
--- a/drivers/net/ethernet/marvell/mvpp2/mvpp2_prs.c
+++ b/drivers/net/ethernet/marvell/mvpp2/mvpp2_prs.c
@@ -312,7 +312,8 @@ static void mvpp2_prs_sram_shift_set(struct mvpp2_prs_entry *pe, int shift,
 	}
 
 	/* Set value */
-	pe->sram[MVPP2_BIT_TO_WORD(MVPP2_PRS_SRAM_SHIFT_OFFS)] = shift & MVPP2_PRS_SRAM_SHIFT_MASK;
+	pe->sram[MVPP2_BIT_TO_WORD(MVPP2_PRS_SRAM_SHIFT_OFFS)] |=
+		shift & MVPP2_PRS_SRAM_SHIFT_MASK;
 
 	/* Reset and set operation */
 	mvpp2_prs_sram_bits_clear(pe, MVPP2_PRS_SRAM_OP_SEL_SHIFT_OFFS,
-- 
2.17.1

